js/app.js

```diff
@@ -11,29 +11,146 @@ Calculator.prototype.getHistoryAsString = function() {
+    let a = this._toNumberOrNull(num1);
+    let b = this._toNumberOrNull(num2);
+    if (a === null || b === null) {
+        alert('Nieprawidłowe liczby. Operacja mnożenia przerwana.');
+        return null;
+    }
+    let result = a * b;
+    const record = a + ' * ' + b + ' = ' + result;
+    this.history.push(record);
+    alert(record);
+    return result;
+}
+
+Calculator.prototype.divide = function(num1, num2) {
+    let a = this._toNumberOrNull(num1);
+    let b = this._toNumberOrNull(num2);
+    if (a === null || b === null) {
+        alert('Nieprawidłowe liczby. Operacja dzielenia przerwana.');
+        return null;
+    }
+    if (b === 0) {
+        alert('Błąd: próbujesz dzielić przez 0. Operacja przerwana.');
+        return null;
+    }
+    let result = a / b;
+    const record = a + ' / ' + b + ' = ' + result;
+    this.history.push(record);
+    alert(record);
+    return result;
 }
 
+Calculator.prototype.pow = function(base, exponent) {
+    let a = this._toNumberOrNull(base);
+    let b = this._toNumberOrNull(exponent);
+
+    if (a === null || b === null) {
+        alert('Nieprawidłowe liczby. Operacja potęgowania przerwana.');
+        return null;
+    }
+
+    if (!Number.isInteger(b)) {
+        alert('Wykładnik musi być liczbą całkowitą (w tej wersji). Operacja przerwana.');
+        return null;
+    }
+
+    if (a === 0 && b < 0) {
+        alert('Błąd: 0 podniesione do ujemnej potęgi jest niedozwolone (dzielenie przez 0).');
+        return null;
+    }
+
+    const absExp = Math.abs(b);
+    let result = 1;
+    for (let i = 0; i < absExp; i++) {
+        result *= a;
+    }
+    if (b < 0) result = 1 / result;
+
+    const record = a + ' ^ ' + b + ' = ' + result;
+    this.history.push(record);
+    alert(record);
+    return result;
+}
+
+/* ————— interakcja z użytkownikiem ————— */
 const calc = new Calculator();
 let action, promptContent, isCorrectAction, number1, number2;
 do { 
-    promptContent = 'Podaj jaką operację chcesz wykonać (+, -, *, /, ^) i potwierdź. \n'; // \n - znak nowej linii
+    promptContent = 'Podaj jaką operację chcesz wykonać (+, -, *, /, ^) i potwierdź. \n';
     promptContent += 'Jeśli chcesz zrezygnować wciśnij Anuluj. \n';
     promptContent += 'Lista poprzednich operacji: \n' + calc.getHistoryAsString();
 
     action = prompt(promptContent);
     isCorrectAction = calc.isCorrectAction(action);
+
     if(isCorrectAction) {
         number1 = prompt('Podaj liczbę nr 1');
+        if (number1 === null) break;
+
         number2 = prompt('Podaj liczbę nr 2');
+        if (number2 === null) break;
 
-        if(action === '+') {
-            calc.add(number1, number2);
+        switch(action) {
+            case '+':
+                calc.add(number1, number2);
+                break;
+            case '-':
+                calc.subtract(number1, number2);
+                break;
+            case '*':
+                calc.multiply(number1, number2);
+                break;
```

Comment:
Jeśli operacji będzie dużo to ta lista będzie bardzo długa. Może warto spróbować napisać to inaczej.
Tutaj znajdziesz podpowiedź: https://devmentor.pl/b/mniej-instrukcji-warunkowych